man/pabrowse.1
.TH pabrowse 1 User Manuals
.SH NAME
pabrowse \- List PulseAudio sound servers on the network
.SH SYNOPSIS
\fBpabrowse
\f1
.SH DESCRIPTION
\fIpabrowse\f1 lists all PulseAudio sound servers on the local network that are being announced with Zeroconf/Avahi.

This program takes no command line arguments.
.SH AUTHORS
The PulseAudio Developers <mzchyfrnhqvb (at) 0pointer (dot) net>; PulseAudio is available from \fBhttp://pulseaudio.org/\f1
.SH SEE ALSO
\fBpulseaudio(1)\f1, \fBavahi-daemon(8)\f1


ChangeLog
2007-11-21 23:09  lennart

	* configure.ac: bump version and soname

2007-11-21 23:03  lennart

	* man/pacat.1.xml.in, man/pulse-daemon.conf.5.xml.in: update man
	  pages a bit

2007-11-21 22:55  lennart

	* src/pulsecore/protocol-native.c: when speaking to a client with a
	  version < 12, hide S32 sample specs, and make them appaear as
	  FLOAT32

2007-11-21 01:31  lennart

	* configure.ac: increment api and protocol version

2007-11-21 01:30  lennart

	* src/daemon/daemon-conf.c, src/daemon/daemon-conf.h,
	  src/daemon/daemon.conf.in, src/daemon/main.c, src/map-file,
	  src/modules/module-suspend-on-idle.c,
	  src/modules/module-volume-restore.c, src/pulse/context.c,
	  src/pulse/def.h, src/pulse/internal.h, src/pulse/operation.c,
	  src/pulse/stream.c, src/pulse/stream.h, src/pulsecore/cli-text.c,
	  src/pulsecore/core.c, src/pulsecore/core.h, src/pulsecore/pid.c,
	  src/pulsecore/pid.h, src/pulsecore/protocol-native.c,
	  src/pulsecore/sink-input.c, src/pulsecore/sink-input.h,
	  src/pulsecore/sink.c, src/pulsecore/source-output.c,
	  src/pulsecore/source-output.h, src/pulsecore/source.c,
	  src/utils/pacat.c, src/utils/pacmd.c: - Check process name when
	  dealing with PID files
	  - Add new PA_STREAM_FIX_CHANNELS, FIX_RATE, FIX_FORMAT,
	  DONT_MOVE, VARIABLE_RATES to pa_sream_flags_t adn implement it
	  - Expose those flags in pacat
	  - Add notifications about device suspend/resume to the protocol
	  and expose them in libpulse
	  - Allow changing of buffer_attr during playback
	  - allow disabling for remixing globally
	  - hookup polkit support

2007-11-21 01:21  lennart

	* src/modules/module-oss.c: minor typo fix

2007-11-21 01:21  lennart

	* PROTOCOL: add short version history of the PA protocol

2007-11-21 01:20  lennart

	* src/pulsecore/native-common.h: add a couple of new opcodes, and
	  document the versions the opcodes where added

2007-11-21 01:19  lennart

	* src/pulsecore/memblockq.c, src/pulsecore/memblockq.h: add API to
	  allow runtime reconfiguration of memblockqs

2007-11-16 23:48  lennart

	* configure.ac: detect whether PolicyKit support is available

2007-11-16 23:47  lennart

	* src/Makefile.am: build PolicyKit support

2007-11-16 23:46  lennart

	* src/daemon/PulseAudio.policy, src/daemon/polkit.c,
	  src/daemon/polkit.h: add interface to PolicyKit

2007-11-14 16:11  lennart

	* src/pulse/thread-mainloop.c: use a prio inheriting mutex for the
	  threaded mainloop, to ease writing of RT clients

2007-11-14 16:11  lennart

	* src/pulse/operation.c: use a free list for allocation
	  pa_operation objects

2007-11-14 16:10  lennart

	* src/pulsecore/pdispatch.c: use a free list for allocating
	  reply_info structs

2007-11-14 16:09  lennart

	* src/modules/module-jack-sink.c, src/modules/module-jack-source.c:
	  add missing pa_boolization

2007-11-13 23:42  lennart

	* src/modules/alsa-util.c, src/modules/alsa-util.h,
	  src/modules/module-alsa-sink.c, src/modules/module-alsa-source.c:
	  Rework ALSA mixer channel detection code. This time we actually
	  care about the channel names the ALSA mixer exports for us

2007-11-13 23:41  lennart

	* src/pulse/channelmap.c: add array size to increase chance of
	  detecting missing updates

2007-11-13 19:56  lennart

	* src/Makefile.am: don't fail if the bt-proximity-helper is not
	  built

2007-11-13 19:39  lennart

	* src/daemon/daemon-conf.c, src/daemon/daemon.conf.in: don't touch
	  RLIMIT:MEMBLOCK by default. This should improve out-of-the-box
	  comaptibility with JACK

2007-11-13 19:21  lennart

	* src/utils/pacmd.c: increase the pacmd timeout a bit

2007-11-13 17:37  lennart

	* src/modules/alsa-util.c, src/modules/alsa-util.h,
	  src/modules/module-alsa-sink.c, src/modules/module-alsa-source.c,
	  src/modules/module-bt-proximity.c, src/modules/module-cli.c,
	  src/modules/module-detect.c, src/modules/module-hal-detect.c,
	  src/modules/module-oss.c, src/modules/rtp/module-rtp-send.c,
	  src/pulsecore/modargs.c, src/pulsecore/modargs.h,
	  src/pulsecore/protocol-esound.c, src/pulsecore/protocol-native.c,
	  src/pulsecore/protocol-simple.c: Completely rework ALSA device
	  selection code: choose the device to open depending on the
	  requested number of channels and channel map. In most cases it
	  will now suffice to set default-channels=6 to enable 5.1 sound
	  for all devices that support it

2007-11-13 17:35  lennart

	* src/pulsecore/module.c: fix loading of load-once modules if no
	  other modules was loaded before

2007-11-11 23:18  lennart

	* configure.ac, src, src/Makefile.am,
	  src/modules/bt-proximity-helper.c,
	  src/modules/module-bt-proximity.c: add new fun module that
	  automatically mutes your audio devices when you leave with your
	  bluetooth phone, and unmutes when you come back

2007-11-11 23:01  lennart

	* src/modules/gconf/module-gconf.c: Port module-gconf to make use
	  of the new API pa_start_child_for_read()

2007-11-11 23:00  lennart

	* src/pulsecore/start-child.c, src/pulsecore/start-child.h: pull
	  code for starting helper processes out of module-gconf, clean it
	  up, and stick into a new API pa_start_child_for_read()

2007-11-11 22:59  lennart

	* src/pulsecore/macro.h: add new function pa_strnull() to simplify
	  passing null strings to non-linux printf()

2007-11-11 02:30  lennart

	* src, src/Makefile.am, src/pulsecore/resampler.c,
	  src/pulsecore/resampler.h, src/pulsecore/sink-input.c,
	  src/pulsecore/source-output.c, src/tests/remix-test.c,
	  src/tests/resampler-test.c: add a simple fully-automatic
	  fully-linearupmixer/downmixer and enable it by default

2007-11-09 18:25  lennart

	* src/daemon/dumpmodules.c, src/modules/gconf/module-gconf.c,
	  src/modules/module-alsa-sink.c, src/modules/module-alsa-source.c,
	  src/modules/module-cli.c, src/modules/module-combine.c,
	  src/modules/module-default-device-restore.c,
	  src/modules/module-defs.h.m4, src/modules/module-detect.c,
	  src/modules/module-esound-compat-spawnfd.c,

src/utils/pabrowse.c
/* $Id: pabrowse.c 1071 2006-07-13 17:35:10Z lennart $ */

/***
  This file is part of PulseAudio.
 
  PulseAudio is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published
  by the Free Software Foundation; either version 2 of the License,
  or (at your option) any later version.
 
  PulseAudio is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.
 
  You should have received a copy of the GNU Lesser General Public License
  along with PulseAudio; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
  USA.
***/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <assert.h>
#include <signal.h>

#include <pulse/pulseaudio.h>
#include <pulse/browser.h>

static void exit_signal_callback(pa_mainloop_api*m, pa_signal_event *e, int sig, void *userdata) {
    fprintf(stderr, "Got signal, exiting\n");
    m->quit(m, 0);
}

static void dump_server(const pa_browse_info *i) {
    char t[16];

    if (i->cookie)
        snprintf(t, sizeof(t), "0x%08x", *i->cookie);
    
    printf("server: %s\n"
           "server-version: %s\n"
           "user-name: %s\n"
           "fqdn: %s\n"
           "cookie: %s\n",
           i->server,
           i->server_version ? i->server_version : "n/a",
           i->user_name ? i->user_name : "n/a",
           i->fqdn ? i->fqdn : "n/a",
           i->cookie ? t : "n/a");
}

static void dump_device(const pa_browse_info *i) {
    char ss[PA_SAMPLE_SPEC_SNPRINT_MAX];

    if (i->sample_spec)
        pa_sample_spec_snprint(ss, sizeof(ss), i->sample_spec);

    printf("device: %s\n"
           "description: %s\n"
           "sample spec: %s\n",
           i->device,
           i->description ? i->description : "n/a",
           i->sample_spec ? ss : "n/a");
           
}

static void browser_callback(pa_browser *b, pa_browse_opcode_t c, const pa_browse_info *i, void *userdata) {
    assert(b && i);

    switch (c) {

        case PA_BROWSE_NEW_SERVER:
            printf("\n=> new server <%s>\n", i->name);
            dump_server(i);
            break;

        case PA_BROWSE_NEW_SINK:
            printf("\n=> new sink <%s>\n", i->name);
            dump_server(i);
            dump_device(i);
            break;

        case PA_BROWSE_NEW_SOURCE:
            printf("\n=> new source <%s>\n", i->name);
            dump_server(i);
            dump_device(i);
            break;
            
        case PA_BROWSE_REMOVE_SERVER:
            printf("\n=> removed server <%s>\n", i->name);
            break;

        case PA_BROWSE_REMOVE_SINK:
            printf("\n=> removed sink <%s>\n", i->name);
            break;

        case PA_BROWSE_REMOVE_SOURCE:
            printf("\n=> removed source <%s>\n", i->name);
            break;

        default:
            ;
    }
}

static void error_callback(pa_browser *b, const char *s, void *userdata) {
    pa_mainloop_api*m = userdata;
    
    fprintf(stderr, "Failure: %s\n", s);
    m->quit(m, 1);
}

int main(int argc, char *argv[]) {
    pa_mainloop *mainloop = NULL;
    pa_browser *browser = NULL;
    int ret = 1, r;
    const char *s;

    if (!(mainloop = pa_mainloop_new()))
        goto finish;

    r = pa_signal_init(pa_mainloop_get_api(mainloop));
    assert(r == 0);
    pa_signal_new(SIGINT, exit_signal_callback, NULL);
    pa_signal_new(SIGTERM, exit_signal_callback, NULL);
    signal(SIGPIPE, SIG_IGN);
    
    if (!(browser = pa_browser_new_full(pa_mainloop_get_api(mainloop), PA_BROWSE_FOR_SERVERS|PA_BROWSE_FOR_SINKS|PA_BROWSE_FOR_SOURCES, &s))) {
        fprintf(stderr, "pa_browse_new_full(): %s\n", s);
        goto finish;
    }

    pa_browser_set_callback(browser, browser_callback, NULL);
    pa_browser_set_error_callback(browser, error_callback, pa_mainloop_get_api(mainloop));
    
    ret = 0;
    pa_mainloop_run(mainloop, &ret);

finish:

    if (browser)
        pa_browser_unref(browser);

    if (mainloop) {
        pa_signal_done();
        pa_mainloop_free(mainloop);
    }

    return ret;
}
